### PR TITLE
Migrate Groovy DSL extension project

### PR DESCRIPTION
**Pull request by artembilan**
* Upgrade to Groovy `4.0.5`
* Apply `groovy` plugin for Groovy module to be able to compile Groovy classes for DSL
* Document new feature

### CODE REVIEW

File: src/reference/asciidoc/groovy-dsl.adoc [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+@Bean
+IntegrationFlow oddFlow() {
+    { IntegrationFlowDefinition flow ->
+	    flow.handle(Object, { p, h -> 'odd' })
+    }
+}
+----
+====
+
+In this case Groovy understands that the closure should be translated into an `IntegrationFlow` anonymous instance and target Java DSL processor parses this construction properly into Java objects.
```
Comment (garyrussell) [inline src/reference/asciidoc/groovy-dsl.adoc:24]:
```suggestion
In this case Groovy understands that the closure should be translated into an `IntegrationFlow` anonymous instance and the target Java DSL processor parses this construction properly into Java objects.
```

File: src/reference/asciidoc/groovy-dsl.adoc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+}
+----
+====
+
+Such a global `integrationFlow()` function expects a closure in builder style for a `GroovyIntegrationFlowDefinition` (a Groovy wrapper for the `IntegrationFlowDefinition`) and produces a regular `IntegrationFlow` lambda implementation.
+See more overloaded `integrationFlow()` variants below.
+
+Many other scenarios require an `IntegrationFlow` to be started from source of data (e.g. `JdbcPollingChannelAdapter`, `JmsInboundGateway` or just an existing `MessageChannel`).
+For this purpose Spring Integration Java DSL provides an `IntegrationFlows` factory with its bunch of overloaded `from()` methods.
```
Comment (garyrussell) [inline src/reference/asciidoc/groovy-dsl.adoc:49]:
```suggestion
For this purpose, Spring Integration Java DSL provides an `IntegrationFlows` factory with a number of overloaded `from()` methods.
```

File: src/reference/asciidoc/groovy-dsl.adoc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+flowFromSupplier() {
+    IntegrationFlows.fromSupplier({ 'bar' }) { e -> e.poller { p -> p.fixedDelay(10).maxMessagesPerPoll(1) } }
+            .channel({ c -> c.queue('fromSupplierQueue') } as Function)
+            .get()
+}
+----
+====
+
+But unfortunately not all `from()` methods are compatible with Groovy structures.
+To fix a gap, Spring Integration provides a Groovy DSL around an `IntegrationFlows` factory.
```
Comment (garyrussell) [inline src/reference/asciidoc/groovy-dsl.adoc:65]:
```suggestion
To solve this, Spring Integration provides a Groovy DSL factory around the `IntegrationFlows` factory.
```

File: src/reference/asciidoc/groovy-dsl.adoc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+It is done as a set of overloaded `integrationFlow()` functions.
+With a consumer for a `GroovyIntegrationFlowDefinition` to declare the rest of the flow as an `IntegrationFlow` closure to reuse the mentioned above experience and also avoid `get()` call in the end.
```
Comment (garyrussell) [inline src/reference/asciidoc/groovy-dsl.adoc:67]:
```suggestion
It is implemented as a set of overloaded `integrationFlow()` functions.
With a consumer for a `GroovyIntegrationFlowDefinition` to declare the remainder of the flow as an `IntegrationFlow` closure to reuse the mentioned above experience and also avoid the need for a `get()` call in the end.
```

**Review by garyrussell — CHANGES_REQUESTED** [review]
Just doc polishing.

**Comment by artembilan** [comment]
Do not merge yet: I see some other minor issues.

**Review by garyrussell — APPROVED** [review]

**Pull request merged**